components/script/dom/byteteeunderlyingsource.rs

```diff
@@ -0,0 +1,457 @@
+}
+
+impl ByteTeeUnderlyingSource {
+    #[allow(clippy::too_many_arguments)]
+    #[allow(clippy::redundant_allocation)]
+    #[cfg_attr(crown, allow(crown::unrooted_must_root))]
+    pub(crate) fn new(
+        reader: Rc<RefCell<ReaderType>>,
+        stream: &ReadableStream,
+        reading: Rc<Cell<bool>>,
+        read_again_for_branch_1: Rc<Cell<bool>>,
+        read_again_for_branch_2: Rc<Cell<bool>>,
+        canceled_1: Rc<Cell<bool>>,
+        canceled_2: Rc<Cell<bool>>,
+        reason_1: Rc<Box<Heap<Value>>>,
+        reason_2: Rc<Box<Heap<Value>>>,
+        cancel_promise: Rc<Promise>,
+        tee_cancel_algorithm: ByteTeeCancelAlgorithm,
+        byte_tee_pull_algorithm: ByteTeePullAlgorithm,
+        can_gc: CanGc,
+    ) -> DomRoot<ByteTeeUnderlyingSource> {
+        reflect_dom_object(
+            Box::new(ByteTeeUnderlyingSource {
+                reflector_: Reflector::new(),
+                reader,
+                stream: Dom::from_ref(stream),
+                branch_1: MutNullableDom::new(None),
+                branch_2: MutNullableDom::new(None),
+                read_again_for_branch_1,
+                read_again_for_branch_2,
+                reading,
+                canceled_1,
+                canceled_2,
+                reason_1,
+                reason_2,
+                cancel_promise,
+                tee_cancel_algorithm,
+                byte_tee_pull_algorithm,
+            }),
+            &*stream.global(),
+            can_gc,
+        )
+    }
+
+    pub(crate) fn set_branch_1(&self, stream: &ReadableStream) {
+        self.branch_1.set(Some(stream));
+    }
+
+    pub(crate) fn set_branch_2(&self, stream: &ReadableStream) {
+        self.branch_2.set(Some(stream));
+    }
+
+    pub(crate) fn forward_reader_error(&self, this_reader: Rc<RefCell<ReaderType>>, can_gc: CanGc) {
+        let this_reader = this_reader.borrow_mut();
+        match &*this_reader {
+            ReaderType::Default(_reader) => {
+                // TODO
+            },
+            ReaderType::BYOB(reader) => {
+                // Upon rejection of thisReader.[[closedPromise]] with reason r,
+                reader
+                    .get()
+                    .expect("Reader should be set.")
+                    .byte_tee_append_native_handler_to_closed_promise(
+                        &self.branch_1.get().expect("Branch 1 should be set."),
+                        &self.branch_2.get().expect("Branch 2 should be set."),
+                        self.canceled_1.clone(),
+                        self.canceled_2.clone(),
+                        self.cancel_promise.clone(),
+                        can_gc,
+                    );
+            },
+        }
+    }
+
+    pub(crate) fn pull_with_default_reader(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        can_gc: CanGc,
+    ) {
+        let mut reader = self.reader.borrow_mut();
+        match &*reader {
```

Comment:
This doesn't seem correct: I think it should be an `if &*reader == eaderType::BYOB(byte_reader)`, followed by "Let readRequest be a [read request](https://streams.spec.whatwg.org/#read-request) with the following [items](https://infra.spec.whatwg.org/#struct-item):" that should run for both reader types.